torchx/schedulers/ray/ray_driver.py

```diff
@@ -77,77 +102,189 @@ def load_actor_json(filename: str) -> List[RayActor]:
         return actors
 
 
-def create_placement_group(replicas: List[RayActor]) -> PlacementGroup:
+def create_placement_group_async(replicas: List[RayActor]) -> PlacementGroup:
+    """return a placement group reference, the corresponding placement group could be scheduled or pending"""
     bundles = []
     for replica in replicas:
         bundles.append({"CPU": replica.num_cpus, "GPU": replica.num_gpus})
 
-    # To change the strategy type
-    # refer to available options here https://docs.ray.io/en/latest/placement-group.html#pgroup-strategy
     pg = ray.util.placement_group(bundles, strategy="SPREAD")
+    return pg
+
+
+@dataclass
+class ActorInfo:
+    """Used to store the information for restoring a failed command actor"""
+
+    pg: PlacementGroup
+    replica: RayActor
+    actor: CommandActor
 
-    _logger.info("Waiting for placement group to start.")
-    ready = pg.wait(timeout_seconds=100)
 
-    if not ready:  # pragma: no cover
-        raise TimeoutError(
-            "Placement group creation timed out. Make sure "
-            "your cluster either has enough resources or use "
-            "an autoscaling cluster. Current resources "
-            "available: {}, resources requested by the "
-            "placement group: {}".format(ray.available_resources(), pg.bundle_specs)
+class RayDriver:
+    def __init__(self, replicas: List[RayActor]) -> None:
+        self.replicas = replicas
+        self.master_node_id: Optional[str] = None  # the actor id of the master node
+        self.rank_0_address: Optional[str] = None
+        self.rank_0_port: Optional[int] = None
+        self.min_nnodes: Optional[int] = replicas[0].min_nnodes
+        self.max_nnodes: int = len(replicas)
+        if self.min_nnodes is None:
+            self.min_nnodes = self.max_nnodes
+
+        self.placement_groups: List[
+            PlacementGroup
+        ] = []  # all the placement groups, shall never change
+        self.actor_info_of_id: Dict[
+            str, ActorInfo
+        ] = {}  # store the info used to recover an actor
+        self.active_tasks: List["ray.ObjectRef"] = []  # list of active tasks
+
+        self.need_more_actors: bool = (
+            True  # need more actors before any actor finishes and exits
         )
-    return pg
+        self.command_actors_count: int = 0  # number of created command actors
 
+    def init_placement_groups(self) -> None:
+        """Initialize all placement groups needed for this job"""
+        # find the actor specifications of a given placement group
+        replica_ix_of_pg: List[int] = [0] + list(
+            range(self.min_nnodes, self.max_nnodes + 1)  # pyre-ignore
```

Comment:
why is there a pyre ignore?

should at least be a pyre-fixme with the error message